### ISSUE TITLE
Replace skyCountsPerSec function with calls to the API in LSST photUtils.

### ISSUE THREAD

**Issue by cwwalter** [LSSTDESC/imSim#45]
Issue #5 implements the ESO sky model.

We should replace the function that converts the sky level electron counts (from @dkirkby and the WL deblending package) with calls to the system utilities in photUtils.  @danielsf has summarized how these functions work in a recent DESC SSim meeting.

The function to be replaced is:

def skyCountsPerSec(surface_brightness=21, filter_band='r',
                    effective_area=32.4*u.m**2, pixel_size=0.2*u.arcsec):

The zero points in this function now are hardcoded from the calculations by @dkirkby and @fjaviersanchez in the "[skyModel_params]" section of data/default_imsim_configs.

**Comment by cwwalter** [comment]
@danielsf can probably point anyone working on this to the proper calls.

**Comment by danielsf** [comment]
I suspect this method is what you are talking about (?)

https://github.com/lsst/sims_photUtils/blob/master/python/lsst/sims/photUtils/SignalToNoise.py#L133

**Comment by cwwalter** [comment]
> I suspect this method is what you are talking about (?)

I don't think so.. those functions calculate the number of electrons given a background for detecting a signal at 5 sigma over the background correct?

I am talking about the general function here:

https://github.com/LSSTDESC/imSim/blob/master/python/desc/imsim/skyModel.py#L23

which comes from the WL deblending package from work from @dkirkby and @fjaviersanchez .  Given the magnitude of the brightness of the sky from the ESO skymodel code from @yoachim in 
lsst.sims.skybrightness  this function calculates the number of electrons per second in a band given the zero point and effective area.

Currently these values are hard coded (and are only there for the r-band).  We need to replace this with the corresponding LSST system values/functions.

Also note how it is used here:

https://github.com/LSSTDESC/imSim/blob/master/python/desc/imsim/skyModel.py#L100

and the changes that will need to be made if we go to multiple snaps.

**Comment by danielsf** [comment]
The method I pointed to takes an m5 level and tells you how many sky counts per pixel corresponds to that m5 level (so: if you just wanted to use OpSim's m5 value as Truth, you could turn that into a sky counts level, you could use this method).

That being said, I forgot that we wanted to use Peter's model.  I imagine that should be possible. I haven't looked at it in detail, though.

**Comment by fjaviersanchez** [comment]
I think that the function that we would have to call is this:

https://github.com/lsst-pst/syseng_throughputs/blob/master/python/calcM5.py#L59-L60

**Comment by ellongley** [comment]
Hello, I was working on rewriting this function, but had a question.  The functions linked to above take an Sed as an argument, either by generating a flat Sed in the case of the function in SingaltoNoise.py or in the case of the function in calcM5.py it reads in a "darksky.dat" file in lsst-pst/syseng_throughputs/darksky.dat, below are the lines:

`darksky = Sed()`

`darksky.readSED_flambda(os.path.join(getPackageDir('syseng_throughputs'),
                                         'siteProperties', 'darksky.dat'))`

`skyCounts[f] = (darksky.calcADU(hardware[f], photParams=photParams)
                        * photParams.platescale**2)`

Does it make sense to use this file for the model that we want to use? 

thank you !

**Comment by danielsf** [comment]
I believe we will want to replace `darksky.dat` with an SED generated by the `sims_skybrightness` software package

https://github.com/lsst/sims_skybrightness

Look at this class

https://github.com/lsst/sims_skybrightness/blob/master/python/lsst/sims/skybrightness/skyModel.py#L134

**Comment by ellongley** [comment]
Hi there ! I still just have version 12 at NERSC, @jchiang87 , @heather999  how should I setup the latest version ? Thank you so much !

**Comment by jchiang87** [comment]
@ellongley From bash, do
```
$ source /global/common/cori/contrib/lsst/lsstDM/setupStack-w_2017_26.sh
```
and that will get you a version of the Stack from week 26, which should be sufficiently recent for this.

**Comment by ellongley** [comment]
great thanks !

**Comment by cwwalter** [comment]
Recording this here: One suggestion from @jchiang87 is that we use this function to calculate the values in an external program and then periodically update the config file values.  The advantage of this is that you aren't recalculating a number that never changes over and over and could easily do overrides.

The disadvantage is that if the system throughputs change you won't automatically get updated zero points.  We would need a procedure to make sure they are kept up-to-date.   For now just make the function work as originally envisioned and then we can decide exactly how to utilize it later.

**Comment by danielsf** [comment]
Would it be possible to store a cached value in the `skyCountsPerSec()` method.  We could initialize the cache to `None`.  When you call `skyCountsPerSec()`, it would check the cache.  If the cached value was `None`, it would do the calculation.  If not, it would return the pre-calculated value.  This would depend on how fine the grid in time, RA, and Dec is (I assume these values are gridded on time, RA, and Dec).

**Comment by cwwalter** [comment]
> This would depend on how fine the grid in time, RA, and Dec is (I assume these values are gridded on time, RA, and Dec

Since this is just the dark sky value we are calculating a zero point I think it shouldn't change with position or time.

**Comment by ellongley** [comment]
Closed by pull request #75.
